zkevm-circuits/src/circuit_tools.rs

```diff
@@ -0,0 +1,953 @@
+        }
+    }};
+
+    // Lookup using a tuple
+    ($cb:expr, ($($v:expr),+) => @$tag:expr) => {{
+        $cb.lookup(
+            concat_with_preamble!(
+                "(",
+                $(
+                    stringify!($v),
+                    ", ",
+                )*
+                ") => @",
+                stringify!($tag),
+            ),
+            $tag.to_string(),
+            vec![$($v.expr(),)*],
+        );
+    }};
+    ($cb:expr, $descr:expr, ($($v:expr),+)  => @$tag:expr) => {{
+        $cb.lookup(
+            Box::leak($descr.into_boxed_str()),
+            $tag.to_string(),
+            vec![$($v.expr(),)*],
+        );
+    }};
+
+    // Lookup using an array
+    ($cb:expr, $values:expr => @$tag:expr) => {{
+        $cb.lookup(
+            concat_with_preamble!(
+                stringify!($values),
+                " => @",
+                stringify!($tag),
+            ),
+            $tag.to_string(),
+            $values.clone(),
+        );
+    }};
+    ($cb:expr, $descr:expr, $values:expr => @$tag:expr) => {{
+        $cb.lookup(
+            Box::leak($descr.to_string().into_boxed_str()),
+            $tag.to_string(),
+            $values.clone(),
+        );
+    }};
+
+    // Put values in a lookup table using a tuple
+    ($cb:expr, @$tag:expr => ($($v:expr),+)) => {{
+        $cb.lookup_table(
+            concat_with_preamble!(
+                "@",
+                stringify!($tag),
+                " => (",
+                $(
+                    stringify!($v),
+                    ", ",
+                )*
+                ")",
+            ),
+            $tag.to_string(),
+            vec![$($v.expr(),)*],
+        );
+    }};
+    // Put values in a lookup table using an array
+    ($cb:expr, @$tag:expr => $values:expr) => {{
+        $cb.lookup_table(
+            concat_with_preamble!(
+                "@",
+                stringify!($tag),
+                " => (",
+                stringify!($values),
+                ")",
+            ),
+            $tag.to_string(),
+            $values,
+        );
+    }};
+}
+
+/// matchx
+#[macro_export]
+macro_rules! _matchx {
+    ($cb:expr, $($condition:expr => $when:expr),* $(, _ => $catch_all:expr)? $(,)?)  => {{
+        let mut conditions = Vec::new();
+        let mut cases = Vec::new();
+        $(
+            $cb.push_condition($condition.expr());
+            let ret = $when.clone();
+            $cb.pop_condition();
+            cases.push(($condition.expr(), ret));
+            conditions.push($condition.expr());
+        )*
+
+        $(
+            let catch_all_condition = not::expr(sum::expr(&conditions));
+            $cb.push_condition(catch_all_condition.expr());
+            let ret = $catch_all;
+            $cb.pop_condition();
+            cases.push((catch_all_condition.expr(), ret));
+            conditions.push(catch_all_condition.expr());
+        )*
+
+        _require!($cb, sum::expr(&conditions) => 1);
```

Comment:
Probably we need to check that the conditions are all boolean?

Reply:
I was hesitant to add this check by default, but will add this because indeed no good reason not to do so (we can always change it so it can be turned off).